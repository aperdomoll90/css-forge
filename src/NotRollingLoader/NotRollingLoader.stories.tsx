import { Meta, StoryFn } from '@storybook/react'
import { NotRollingLoader } from './NotRollingLoader'
import StorybookContainer from '../StorybookContainer'
import { NotRollingLoaderPropsType } from './NotRollingLoader.types'

export default {
  title: 'Loaders',
  component: NotRollingLoader,
  parameters: {
    jest: ['NotRollingLoader.test.tsx'],
  },
  argTypes: {
    tracks: { options: [2, 4, 7, 8], control: { type: 'radio' } },
  },
} as Meta

const NotRollingLoaderTemplate: StoryFn<NotRollingLoaderPropsType> = args => (
  <StorybookContainer>
    <NotRollingLoader {...args} />
  </StorybookContainer>
)

export const NotRollingLoaderComponent = NotRollingLoaderTemplate.bind({})
NotRollingLoaderComponent.args = {
  tracks: 10,
}
