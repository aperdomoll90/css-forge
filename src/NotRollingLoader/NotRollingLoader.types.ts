export interface NotRollingLoaderPropsType {
    tracks?: number;
    trackColor?: string;
  }