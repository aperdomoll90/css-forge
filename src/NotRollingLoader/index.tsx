export { NotRollingLoader } from "./NotRollingLoader"
export * from './NotRollingLoader.types'