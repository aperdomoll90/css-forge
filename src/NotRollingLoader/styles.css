.not-rolling-loader-wrapper {
  width: 20rem;
  aspect-ratio: 1;
  border-radius: 50%;
  outline: 2px solid var(--trackColor);
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
}

.not-rolling-loader-track {
  position: absolute;
  width: 100%;
  height: 1px;
  background: var(--trackColor);
  overflow: visible;
  transform: rotate(calc(-180deg / var(--tracks) * var(--index)));
}
.not-rolling-loader-track-inner::before {
  content: '';
  position: absolute;
  top: -1rem;
  left: -1rem;
  width: 2rem;
  aspect-ratio: 1;
  border-radius: 50%;
  background: var(--color);
}

.not-rolling-loader-track-inner {
  width: 100%;
  height: 100%;
  position: relative;
  animation: animateTrack 1s infinite ease-in-out;
  animation-direction: alternate;
  animation-delay: calc(1s / var(--tracks) * var(--index));
}

@keyframes animateTrack {
  0% {
    transform: translateX(0);
  }
  100% {
    transform: translateX(100%);
  }
}
