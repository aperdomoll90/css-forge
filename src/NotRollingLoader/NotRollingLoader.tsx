import { useEffect } from 'react'
import { NotRollingLoaderPropsType } from './NotRollingLoader.types'
import './styles.css'
import { ColorByIndex } from '../utils/ColorManipulation'

export const NotRollingLoader: React.FC<NotRollingLoaderPropsType> = ({ tracks = 8, trackColor = 'transparent' }) => {
  useEffect(() => {
    const addTrack = (index: number) => {
      const track = document.createElement('div')
      track.classList.add('not-rolling-loader-track')
      track.style.setProperty('--index', `${index}`)
      track.style.setProperty('--trackColor', `${trackColor}`)
      track.style.setProperty('--tracks', `${tracks}`)
      track.style.setProperty('--color', ColorByIndex(index, tracks))
      const trackInner = document.createElement('div')
      trackInner.classList.add('not-rolling-loader-track-inner')
      track.appendChild(trackInner)
      const wrapper = document.getElementById('not-rolling-loader-wrapper')
      wrapper && wrapper.appendChild(track)
    }
    for (let i = 0; i < tracks; i++) {
      addTrack(i)
    }
  }, [])

  return <div id='not-rolling-loader-wrapper' className='not-rolling-loader-wrapper' style={{ '--trackColor': trackColor } as React.CSSProperties}></div>
}
