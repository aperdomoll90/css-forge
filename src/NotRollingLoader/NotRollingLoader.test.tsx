import { render } from '@testing-library/react'
import { NotRollingLoader } from './NotRollingLoader'
import { NotRollingLoaderPropsType } from './NotRollingLoader.types'

describe('NotRollingLoader component', () => {
  let props: NotRollingLoaderPropsType = {
    tracks: 8,
  }
  const renderComponent = () => render(<NotRollingLoader {...props} />)
  it('should match snapshots', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })
})
