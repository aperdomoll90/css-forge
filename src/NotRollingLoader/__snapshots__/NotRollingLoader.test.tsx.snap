// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotRollingLoader component should match snapshots 1`] = `
<div>
  <div
    class="not-rolling-loader-wrapper"
    id="not-rolling-loader-wrapper"
    style="--trackColor: transparent;"
  >
    <div
      class="not-rolling-loader-track"
      style="--index: 0; --trackColor: transparent; --tracks: 8; --color: hsl(0, 100%, 50%);"
    >
      <div
        class="not-rolling-loader-track-inner"
      />
    </div>
    <div
      class="not-rolling-loader-track"
      style="--index: 1; --trackColor: transparent; --tracks: 8; --color: hsl(45, 100%, 50%);"
    >
      <div
        class="not-rolling-loader-track-inner"
      />
    </div>
    <div
      class="not-rolling-loader-track"
      style="--index: 2; --trackColor: transparent; --tracks: 8; --color: hsl(90, 100%, 50%);"
    >
      <div
        class="not-rolling-loader-track-inner"
      />
    </div>
    <div
      class="not-rolling-loader-track"
      style="--index: 3; --trackColor: transparent; --tracks: 8; --color: hsl(135, 100%, 50%);"
    >
      <div
        class="not-rolling-loader-track-inner"
      />
    </div>
    <div
      class="not-rolling-loader-track"
      style="--index: 4; --trackColor: transparent; --tracks: 8; --color: hsl(180, 100%, 50%);"
    >
      <div
        class="not-rolling-loader-track-inner"
      />
    </div>
    <div
      class="not-rolling-loader-track"
      style="--index: 5; --trackColor: transparent; --tracks: 8; --color: hsl(225, 100%, 50%);"
    >
      <div
        class="not-rolling-loader-track-inner"
      />
    </div>
    <div
      class="not-rolling-loader-track"
      style="--index: 6; --trackColor: transparent; --tracks: 8; --color: hsl(270, 100%, 50%);"
    >
      <div
        class="not-rolling-loader-track-inner"
      />
    </div>
    <div
      class="not-rolling-loader-track"
      style="--index: 7; --trackColor: transparent; --tracks: 8; --color: hsl(315, 100%, 50%);"
    >
      <div
        class="not-rolling-loader-track-inner"
      />
    </div>
  </div>
</div>
`;
